core/trie/proof.go

```diff
@@ -165,12 +189,328 @@ func VerifyProof(root *felt.Felt, key *Key, value *felt.Felt, proofs []ProofNode
+	}
+	if leafKey.Test(leafKey.len - parentKey.len - 1) {
+		parentNode.Right = childKey
+		parentNode.Left = nilKey
+	} else {
+		parentNode.Right = nilKey
+		parentNode.Left = childKey
+	}
+	return childKey, nil
+}
+
+// ProofToPath returns a set of storage nodes from the root to the end of the proof path.
+// It will contain the hashes of the children along the path, but only the key of the children
+// along the path. The final node must contain the hash of the leaf if the leaf is set.
+// It will not contain the leaf node even if it is set.
+func ProofToPath(proofNodes []ProofNode, leafKey *Key, hashF hashFunc) ([]StorageNode, error) {
+	pathNodes := []StorageNode{}
+
+	// Child keys that can't be derived are set to nilKey, so that we can store the node
+	zeroFeltBytes := new(felt.Felt).Bytes()
+	nilKey := NewKey(0, zeroFeltBytes[:])
+
+	for i, pNode := range proofNodes {
+		// Keep moving along the path (may need to skip nodes that were compressed into the last path node)
+		if i != 0 {
+			if skipNode(pNode, pathNodes, hashF) {
+				continue
+			}
+		}
+
+		var parentKey *Key
+		parentNode := Node{}
+
+		// Set the key of the current node
+		compressParent, compressParentOffset, err := compressNode(i, proofNodes, hashF)
+		if err != nil {
+			return nil, err
+		}
+		parentKey, err = getParentKey(i, compressParentOffset, leafKey, pNode, pathNodes, proofNodes)
+		if err != nil {
+			return nil, err
+		}
+
+		// Don't store leafs along proof paths
+		if parentKey.len == 251 { //nolint:gomnd
+			break
+		}
+
+		// Set the value of the current node
+		parentNode.Value = pNode.Hash(hashF)
+
+		// Set the child key of the current node.
+		childKey, err := assignChild(i, compressParent, &parentNode, &nilKey, leafKey, parentKey, proofNodes, hashF)
+		if err != nil {
+			return nil, err
+		}
+
+		// Set the LeftHash and RightHash values
+		parentNode.LeftHash, parentNode.RightHash = getLeftRightHash(i, proofNodes)
+
+		pathNodes = append(pathNodes, StorageNode{key: parentKey, node: &parentNode})
+
+		// break early since we don't store leafs along proof paths, or if no more nodes exist along the proof paths
+		if childKey.len == 0 || childKey.len == 251 {
+			break
+		}
+	}
+	return pathNodes, nil
+}
+
+func skipNode(pNode ProofNode, pathNodes []StorageNode, hashF hashFunc) bool {
+	lastNode := pathNodes[len(pathNodes)-1].node
+	noLeftMatch, noRightMatch := false, false
+	if lastNode.LeftHash != nil && !pNode.Hash(hashF).Equal(lastNode.LeftHash) {
+		noLeftMatch = true
+	}
+	if lastNode.RightHash != nil && !pNode.Hash(hashF).Equal(lastNode.RightHash) {
+		noRightMatch = true
+	}
+	if noLeftMatch && noRightMatch {
+		return true
+	}
+	return false
+}
+
+func getLeftRightHash(parentInd int, proofNodes []ProofNode) (*felt.Felt, *felt.Felt) {
+	parent := &proofNodes[parentInd]
+	if parent.Binary != nil {
+		parentBinary := parent.Binary
+		return parentBinary.LeftHash, parentBinary.RightHash
+	} else {
+		parentBinary := proofNodes[parentInd+1].Binary
```

Comment:
Don't we need ro check that `proofNodes[parentInd+1].Binary != nil` as we did with the previous node?
also `parentInd+1 < len(proofNodes)` 🤔 

Maybe we should assign the `parent` and have a single return:

```
parent := &proofNodes[parentInd]
if parent.Binary == nil {
    parent := &proofNodes[parentInd+1]
}
return parent.Binary.LeftHash, parent.Binary.RightHash
```

Reply:
> Don't we need ro check that proofNodes[parentInd+1].Binary != nil as we did with the previous node?
Nope, if proofNode[i] is an edge then we know proofNode[i+1] is Binary (it's not possible for an edge node to follow an edge node).

> Maybe we should assign the parent and have a single return
Yup, this is much nicer!

> parentInd+1 < len(proofNodes)
Good point. Let me think about that

Reply:
>    parentInd+1 < len(proofNodes)

I added a check for this